### ISSUE TITLE
ECONNRESET when using fastGet exits my script instead of throwing an error

### ISSUE THREAD

**Issue by kamilmielnik** [theophilusx/ssh2-sftp-client#196]
I have a script that loops over several thousands of files on the FTP. I'm using `fastGet` to download each file. The script runs for several hours. Sometimes `ECONNRESET` is thrown when using `fastGet` in (as it seems) a random moment (2 out of 3 last runs). And it breaks my script.

The problem is that the following code does not work as I'd expect it to:
```js
try {
    await sftpClient.fastGet(filepath, targetFilepath);
} catch (error) {
    console.log("This never executes, but I expect it to.");
    console.log("I'd be happy to put some code to re-establish the FTP connection here.");
}
```

This is the output of the script:
```
/my-script/node_modules/ssh2-sftp-client/src/utils.js:80
    throw formatError(err, name);
    ^

Error: : read ECONNRESET
    at formatError (/my-script/node_modules/ssh2-sftp-client/src/utils.js:54:18)
    at Client.<anonymous> (/my-script/node_modules/ssh2-sftp-client/src/utils.js:80:11)
    at Client.emit (events.js:210:5)
    at Socket.<anonymous> (/my-script/node_modules/ssh2/lib/client.js:307:10)
    at Socket.emit (events.js:215:7)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  code: 'ECONNRESET'
}
```

So I understand that this is what happens: `uncaughtException` event is raised. And as [described in node.js docs](https://nodejs.org/api/all.html#process_event_uncaughtexception):

> The 'uncaughtException' event is emitted when an uncaught JavaScript exception bubbles all the way back to the event loop. By default, Node.js handles such exceptions by printing the stack trace to stderr and exiting with code 1, overriding any previously set process.exitCode

----

I'm not sure whether this is a bug or my expectations about `try`/`catch` block around `fastGet` call are wrong. 

----

edit: I'm using latest version (`4.3.0`)

----

edit 2:
As [node docs state](https://nodejs.org/api/all.html#process_warning_using_uncaughtexception_correctly):
> The correct use of 'uncaughtException' is to perform synchronous cleanup of allocated resources (e.g. file descriptors, handles, etc) before shutting down the process. **It is not safe to resume normal operation after 'uncaughtException'.**
> To restart a crashed application in a more reliable way, whether uncaughtException is emitted or not, an external monitor should be employed in a separate process to detect application failures and recover or restart as needed.

So unfortunately it's not easy to recover...

**Comment by theophilusx** [comment]
The econreset error is caused when the remote sftp server resets the connection unexpectedly. In general, you cannot use try/catch to catch exceptions generated via an event emitter (see the node manual for an explanation). In basic terms, the problem is that the try/catch block may have exited before the error event fires. The only reliable way to deal with emitted exceptions is with a listener. 

The problem is that it appears this error is being emitted by the socket object and the SSH2/SFTPStreams modules do not provide any API to attach event listeners on the low level socket  object.  The ssh2 object does have listeners for both error and uncaughtException events, but it would appear that the SSH2 module is not catching and  re-emitting these events, so the listeners are unable to be of any use. 

So, in short, I'm not sure how to catch these events. It could be this is something which needs to be fixed in the ssh2 module itself. 

I will try to write a very simple script which will just use the ssh2 module and which uses only the available events provided by that module which can be used to download a large number of files. We can then use that script to see if we still cannot catch the exception via the listeners. If this is the case, then we can raise an issue with the ssh2 author. On the other hand, if the ssh2 scirpts is able to deal with it, it means I've got something wrong with the ssh2-sftp-client setup and will need to dig deeper. 

give me a couple of days and I'll get back to you with a scirpt. If you can help by running it to try and generate the error, this will help a lot (one of the problems here is in being able to reliably reporduce the issue). 

The issue here, which I'm not sure how to solve, is that the event is not being raised by the ssh2 or ssh2-streams objects the the ssh2-sftp-client module wraps. I have a listener for uncaught exceptions on the ssh2 object, but it isn't that object which is emitting the error - the error is being emitted by a low level socket module and there is no API provided by the ssh2/sftp-stream modules which would allow me to attached a listener on it to handle the exception. 

What I think may work is to put a listener on the close event.

**Comment by kamilmielnik** [comment]
> give me a couple of days and I'll get back to you with a scirpt. If you can help by running it to try and generate the error, this will help a lot (one of the problems here is in being able to reliably reporduce the issue).

That's great, thank you!
I'll be happy to help to reproduce the error with your script :+1:

**Comment by theophilusx** [comment]
BTW what platform are you running the script on (Linux, max or windows) and do you know what platform the ftp server is running on? 

I seem to see more issues with econreset from windows platform, so just interested as I only see it very rarely on Linux and Mac (and it is typically due tot he remote server becoming unavailable, which also means that even if we can find a way to catch this exception, it will be unlikely your script can do much other than log the error and schedule another attempt at a later time i.e. when the server is again available. 

One other thought - I'm assuming you just make one connection and then download all the files rather than make connection, download file, disconnect, make connection, download file, disconnect etc. Reason being is that most servers have rate limiting on incoming unauthenticated connections. Because SSH needs to do some negotiation in order to establish identity and ciphers etc, there is a period where the initial connection is classified as unauthenticated. Most servers will only allow 10 or so such unauthenticated connections at a time. If you try to make too many connections too close to each other, you will often see econreset errors as the remote server begins to drop the connections. However, it is made a little more complicated - the server will start dropping only a percentage of connections once you pass an initial threshold (lets say 40%) until the number hits a hard cutoff at which time it will start to drop all connections. As a result, you will see what appears to be random econnreset errors to begin with and if you persist trying to connect, eventually all connections being rejected.

**Comment by theophilusx** [comment]
I have created a new branch in the repository called econrest-handling where I've added some logging statements to listeners, change the uncaughtException Listener to simply log uncaught exceptions and added two scripts which might help provide some additional information. 

The first script is examples/bulk-fastget.js, which will just download the same fuile multiple times. I've been running it for 1000+ times using the largish test-file1.txt file in the tests/testData directory. I've not been able to generate errors (even turning off wifi and physically unplogging the network cable doesn't raise an error - the script just hangs until the network comes back and then resumes. I really need to complete writing my sftp server which will enable me to raise errors at will in order to reproduce error conditions in a consistent manner. 

The second script is tools/lowlevel-gastget.js, which does the same thing as the first script, but only uses ssh2 and not ssh2-sftp-client. 

both scripts take 3 command line arguments 

1. Path to the remote file on the remote sftp server
2. path to directory to use to donwload the files into
3. Number of interations to run e.g. 100 or 1000

I use dotenv to store the connection information, so you can put the server, port, username, password into a .env file in the root directory and it should work. Otherwise, just hard code your values into the script. 

If you can run each of these scripts a few times until the errors you have seen get generated, that would be great. You will probably have to do this multiple times. 

Thinking about your overall issue of how to deal with failed connections and the remote server unexpectedly closing/resetting the connection, you could try adding a 'close' or 'end' listener which then takes some action. the underlying sftp module will emit an end and then a close event when the connection is closed. I have a flag variable in the client object called 'endCalled', which the end() method of the client sets to true prior to calling end. This is used to distinguish between 'end' and 'close' events being expected and unexpected (the ssh2 module will always emit an end and a close event, but we are really only interested in those events when they are unexpected i.e. not as a result of calling end(). 

If you have a hard time generating the errors, you could try just running with the econrest-handling branch instead of the production version for a while. I should still work fine, just might generate a little more logging noise.

**Comment by theophilusx** [comment]
I have made some more tweaks in the end() method and the close and end event listeners within the econrest-handling branch which might improve matters. 

Looking at your issue, I think the only way you can reliably deal with the econrest error is to use an event listener on the error event. You cannot reliably manage this with a try/catch as the event can be raised when your execution context is outside any try/catch block. You will need to use an error event handler, which you can add using the on() method. 

I have modified the uncaughtException handler so that if an error event does occur and the default error handler raises an exception, but you are outside the execution context, it will log the error, but not exit the script. This should allow your error handler to run and do whatever action you decide is appropriate when the connection is ended due to an error.

**Comment by kamilmielnik** [comment]
> BTW what platform are you running the script on (Linux, max or windows) and do you know what platform the ftp server is running on?

I'm running it on Linux. I have no idea what platform the FTP servers is running on. 

> One other thought - I'm assuming you just make one connection and then download all the files rather than make connection, download file, disconnect, make connection, download file, disconnect etc.

Yes, your assumption is correct.

> Reason being is that most servers have rate limiting on incoming unauthenticated connections. 

It is an authenticated connection and there were no other connections (at least with the same credentials) at the times when I got these ECONNRESET errors. 

> If you can run each of these scripts a few times until the errors you have seen get generated, that would be great. You will probably have to do this multiple times.

Thank you, I will! But as you have noticed this will take time, so hopefully I will be able to get back to you with some results at the end of this week.

> I have modified the uncaughtException handler so that if an error event does occur and the default error handler raises an exception, but you are outside the execution context, it will log the error, but not exit the script. This should allow your error handler to run and do whatever action you decide is appropriate when the connection is ended due to an error.

So do you think there is no point in attempting to `try/catch` it, and I should just go ahead and implement the error handling logic with the error event before testing out the branch you've prepared?

**Comment by theophilusx** [comment]
>> Reason being is that most servers have rate limiting on incoming unauthenticated connections.
>
> It is an authenticated connection and there were no other connections (at least with the same credentials) at the times when I got these ECONNRESET errors.
>

OK, but just being pedantic - when you do an ssh connection, it is
initially unauthenticated. The client and server need to go through a
'handshake' process before the connection moves from unauthenticated to
authenticated. This is where people get tripped up as they believe that
because they are supplying a username/password it is an authenticated
connection. If all goes well, it will eventually transition to an
authenticated connection, but initially it is unauthenticated. If you
are making connections via a script, it is very easy to reach the
default 10 unauthenticated connections and begin to see connections
reset (I don't believe this is your issue).

You should still use try/catch to catch other errors, but you won't be
able to reliably catch some error events, such as econreset. For those,
you need to use a listener on the client which is triggered by the error
event. The challenge then is what to do when this listener fires as this
will occur asynchronously and not in your main execution path.

With the changes I've made, what should now occur is that either the
event does throw an error while execution context is inside a try/catch,
so all good or the error is caught by the uncaughtException handler and
logged (as opposed from bubbling up and crashing the whole process). The
end event will then fire and set the sftp object to undefined. The next
thing your script tries to do after this should fail with a "No sftp
connection available" error, which will be caught in a try/catch. So,
fingers crossed and things might 'just work' for you now. Just be aware
that the 'no sftp connection available' error may indicate that no
connection was made or an existing connection was lost.

Where things might still come unstuck is if the econrest error is being
raised by a different event object (such as the socket object) and not
being relayed through the ssh2 client object. If this is the case,
listeners added on the ssh2 client object will have no effect (including
the uncaughtException listener) and the script will still crash. If this
is the case, we may need to work with the ssh2 author to find a
solution. Let's see what happens with your testing (I have tried many
times with multiple sftp servers and not been able to generate the right
conditions to test adequately). 

If you want to get more info, you can also add your own listeners for
any of the events raised by the ssh2 client object. See the ssh2 readme
for a list of what events are supported.

--
Tim Cross

**Comment by kamilmielnik** [comment]
Today I ran 2000 successful iterations of both test scripts (`examples/bulk-fastget.js` & `tools/lowlevel-gastget.js`) in my local environment against the FTP server that used to give me `ECONNRESET` error.
I didn't reproduce the error.

The environment where my script runs (and get `ECONNRESET`) is not my local environment though. It'd require effort of other people to allow running aforementioned test scripts in that environment. So all I can do is to update my script's `ssh2-sftp-client` dependency to `econreset-handling` branch. I did that, and the script finished successfully within a few hours but `ECONNRESET` never happened. 

I'll modify my script to needlessly run more FTP operations and I will let the script run overnight. :crossed_fingers: 

I'll post next update in 24h.

**Comment by kamilmielnik** [comment]
Got `ECONNRESET` when using `econreset-handling` branch!

Output:
```
 Error listener fired: read ECONNRESET
/my-script/node_modules/ssh2-sftp-client/src/utils.js:81
    throw formatError(err, name);
    ^

Error: : read ECONNRESET
    at formatError (/my-script/node_modules/ssh2-sftp-client/src/utils.js:54:18)
    at Client.<anonymous> (/my-script/node_modules/ssh2-sftp-client/src/utils.js:81:11)
    at Client.emit (events.js:210:5)
    at Socket.<anonymous> (/my-script/node_modules/ssh2/lib/client.js:307:10)
    at Socket.emit (events.js:215:7)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  code: 'ECONNRESET'
}
```

**Comment by theophilusx** [comment]
OK, disappointing, but useful. I have a couple of other ideas to
try. I'll make some changes to the econrest-handling branch and see if
that improves matters. I'll let you know.



-- 
Tim Cross

**Comment by theophilusx** [comment]
I've made some more changes and pushed them up. This *should* give the
behaviour I referenced previously i.e. no uncaught exception, but after
the error is thrown, all subsequent actions will fail.

My big concern now is that I suspect any download/upload in progress
when such an error occurs may just 'hang' indefinitely. I've seen this
before and it is difficult to solve cleanly. About all we can do is add
a timeout, which I don't like because determining a reasonable default
timeout is difficult.

Give it a try and see if it helps with your issue.


-- 
Tim Cross

**Comment by kamilmielnik** [comment]
> Give it a try and see if it helps with your issue.

Script output:
```
 Error listener fired: read ECONNRESET
: Connection closed unexpectedly
```

>  but after the error is thrown, all subsequent actions will fail.

I understand that all subsequent actions should fail because connection has been closed - that would be great. But this is not the case. My script stopped (with `0` exit code) right after output pasted above.

**Comment by theophilusx** [comment]
That would be the hanging issue I mentioned. All I can think to do is
try and introduce some sort of 'timeout' facility, which won't be
trivial. Alternatively, adding a custom error event listener inside each
Promise might work. either way, it will take some time to implement. 

In the meantime, you could add your own error listener to take
appropriate action when an error event is raised. 
-- 
Tim Cross

**Comment by theophilusx** [comment]
I made some more changes to the econrest-handling branch which I'm
hoping will help address your issue. Please try using the most recent
version and let me know if it captures the error and throws it in a way
you can now catch.

What I have done is add a method specific error handler to every
promise. This should result in the promise being rejected when an error
event is emitted. The main risk is that if I've not ensured the handlers
are removed correctly, we will run into the maxListeners warning. There
is still a possibility that an error may be raised in-between the end of
one promise and before the beginning of another. If this occurs, it will
result in the uncaughtException problem. I could add a default handler
to report this case, but cannot do anything to return an error. 
-- 
Tim Cross

**Comment by kamilmielnik** [comment]
>  Please try using the most recent
version and let me know if it captures the error and throws it in a way
you can now catch.

Yes it did!

I've enhanced my script with logic to attempt (3 times) to re-establish the connection after 60 seconds delay. I'll let you know whether this solved my issue.

**Comment by kamilmielnik** [comment]
Unfortunately my logic didn't handle the `ECONNRESET` properly. I guess I should have read your code first before testing it. But still I believe `ssh2-sftp-client` code should be changed a little.

So the `try`/`catch` block does catch the `ECONNRESET` error now. But that error is being thrown as a `string`, not as an instance of `Error` (which is not considered a good practice).

https://github.com/theophilusx/ssh2-sftp-client/commit/a3d826f8e6be52a0e614294ef62c0c0c076f5841#diff-2b4ca49d4bb0a774c4d4c1672d7aa781R80
Suggested change:
```js
reject(err);
```

I believe after this change lands in `master` (and a new version of `ssh2-sftp-client` is released), this issue can be considered as resolved :) But I'll be happy to test it one more time.

**Comment by theophilusx** [comment]
Agree. That was my mistake. The reject() calls should return the error
object. Will fix and let you know.

-- 
Tim Cross

**Comment by theophilusx** [comment]
I have updated the econrest-handling branch so that reject() now uses
the error object. Will release a new version later today or tomorrow.


-- 
Tim Cross

**Comment by theophilusx** [comment]
I have pushed version 4.3.1 to npmjs.com

**Comment by kamilmielnik** [comment]
I have done some testing over the weekend.

Before you released 4.3.1, I was testing the `econreset-handling` and `ECONNRESET` happened once (when I was downloading a file) - it was caught properly and my script was able to recover from it by re-connecting after 60s :+1: 

But after you've released 4.3.1 and I started using it, I got this:
```
events.js:187
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:201:27)
Emitted 'error' event on Client instance at:
    at Socket.<anonymous> (/my-script/node_modules/ssh2/lib/client.js:307:10)
    at Socket.emit (events.js:215:7)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: 'ECONNRESET',
  code: 'ECONNRESET',
  syscall: 'read',
  level: 'client-socket'
}
```

Judging by the output of my script, this happened when there was no SFTP operation in progress (downloaded file was being processed), but the SFTP connection was open.
So I understand that it means that after all I'm going to need to listen to `ssh2` client error events anyway. Can you please confirm whether my understanding is correct?

**Comment by theophilusx** [comment]
Well that sounds about right. All I did to release 4.3.1 was to merge in
the econreset-handling branch. Depending on how long your processing
takes, I would suspect the sftp connection has timed out and the remote
server has closed it, resulting in the econreset error being raised. As
your not actually doing any other processing at that point, there is no
active promise and no active errorListener (or try/catch around the sftp
connection), so the error bubbles up to the top level and throws the
uncaught exception error with the econrest error object.

There are a few things you could try. Adding your own error listener is
certainly an option. The slightly tricky part might be when an error is
thrown during when both listeners are active (both will fire). As I've
just used .on(), we cannot guarantee order the listeners will
fire. Probably the easiest thing to do would be to check for ECONNRESET
error codes in your try/catch block and ignore them (leave them to be
handled by your error listener instead). 

Another alternative which might work would be to add a uncaughtException
handler to your process (e.g. process.on('uncaughtException', ...)) and
take whatever action seems appropriate, though for some reason, that
feels a little 'hacky' to me :-(.





--
Tim Cross

**Comment by theophilusx** [comment]
Thinking about this some more. something else I could try is to add a 'default' error listener which only takes some action if the other listeners don't fire (i.e. outside one of the promises). All this listener could do is log the error (to console.error) and set the client.sftp object to undefined. This would result in any subsequent attempts to use the connection failing, which could be handled in the 'normal' way.  (it would be nice to have a solution which does not require clients to add an error listener IMO).

**Comment by kamilmielnik** [comment]
>  This would result in any subsequent attempts to use the connection failing, which could be handled in the 'normal' way. (it would be nice to have a solution which does not require clients to add an error listener IMO).

This sounds reasonable to me. It appears that I would not have to modify my script for this, so as a library user I am all up for this change :+1:

**Comment by kamilmielnik** [comment]
Well, whether I'd have to modify my script, depends on how would you implement this. Would this error be thrown as ECONNRESET or as something else?

**Comment by theophilusx** [comment]
> Well, whether I'd have to modify my script, depends on how would you implement this. Would this error be
> thrown as ECONNRESET or as something else?

Short answer yes, it will have ECONNRESET as the error.code

Long answer -

When the library throws an error, if the error it is throwing is the
result of an error the library has caught, I use the same error code
(the error message may be modified to provide more context, but the code
is left alone). The library will set an error code if the error is not
from a caught error (in which case a library specific error code will be
used) or if the caught error does not have an error code (error codes
were only standardised in node v10.x, so some libs still don't set an
error.code property in errors they generate).

So basically, I don't change error codes, but will add one if one does
not exist. All errors thrown by the lib should have an error.code property.


-- 
Tim Cross

**Comment by theophilusx** [comment]
>>  This would result in any subsequent attempts to use the connection failing, which could be handled in the 'normal' way. (it would be nice to have a solution which does not require clients to add an error listener IMO).
>
> This sounds reasonable to me. It appears that I would not have to modify my script for this, so as a library user I am all up for this change :+1: 

OK, lets see if this can work. I will let you know. Note that I was just
starting to add two new features (uploadDir and downloadDir) and have
refactored the code in preparation for v5.0.0. I will make a branch of
there as it will be easier in the long run. will let you know what the
new branch is once done (could take a couple of days as family
commitments are ramping up for Christmas, so a little busy). 

-- 
Tim Cross

**Comment by kamilmielnik** [comment]
>  will let you know what the
new branch is once done (could take a couple of days as family
commitments are ramping up for Christmas, so a little busy).

For the same reasons I will only be able to test it after new year anyway.
Thank you and Merry Christmas :)

**Comment by theophilusx** [comment]
When you get the chance, please try the release-5.0.0 branch.

**Comment by kamilmielnik** [comment]
I tried the `release-5.0.0` branch today and I got this console output:
```
Error event: read ECONNRESET
sftp: Connection closed unexpectedly
```

My script continued from there but it just so happened that no FTP operations followed after this event, so I don't know what would happen with e.g. `fastGet` call. I'll modify my script to ensure this happens because I want to know how my script would behave.

The only call to `ssh2-sftp-client` after this was `close` which interestingly did not fail (should closing a closed connection fail?).

**Comment by theophilusx** [comment]
The fact the script didn't just blow up I will take as a win :-). No, calling end() when the sftp connection has already been closed does not cause an error.

**Comment by kamilmielnik** [comment]
For the record: my previous comment described a case where there was no FTP operation in progress.

But today ECONNRESET occurred while `fastGet` was in progress, and unfortunately try / catch statement didn't catch the error, and the script did not continue from there. 

The output of the script was the same:
```
Error event: read ECONNRESET
sftp: Connection closed unexpectedly
```

It's worth noting that my script exited with 0 exit code.

**Comment by theophilusx** [comment]
That output indicates the default error handler processed the error
event. This should only happen if there is no method error handler
defined. The method error handler is defined once the Promise wrapper
around the ssh2/ssh2-streams method is created (it cannot be created
earlier as it needs access to the Promise's reject method).

Each of the ssh2-sftp-client methods does some error checking prior to
creaton of the promise, which is the last thing the methods do (in most
cases, but different for things like uploadDir/downloadDir or recursive
mkdir).

so only one of two things must have occured for you to get that output

1. The error event fired before fastGet() had setup the Promise and
associated error handler. This would result in the 'default' error
handler being called, which cannot be caught by try/catch.

2. The method error handler is not being added as it should. Can't see
why this wold be, but it is possible. I will need to think about how I
can verify this.

In the meantime, I've added a little bit of console debugging code which
might help provide more information regarding when error handlers
fire. I've updated the release-5.0.0 branch. See if that helps.

If your able to strip down your code to a minimal version which
reproduces the issue, it would be really handy to get a copy. It would
be much easier if I could somehow reproduce this issue. I'm using this
module to download lots of large files every day and only very rarely
see econreset issues. When we have, they have been caught and handled as
expected. It is possible you are doing something I've not thought of and
therefore don't adequately test for it, so the issue is not picked up
with my testing.

**Comment by kamilmielnik** [comment]
> 1. The error event fired before fastGet() had setup the Promise and
associated error handler. This would result in the 'default' error
handler being called, which cannot be caught by try/catch.

But then why would the script exit?

> 2. The method error handler is not being added as it should. Can't see
why this wold be, but it is possible. I will need to think about how I
can verify this.

My hunch is that this is more likely (see above).

----

> I've updated the release-5.0.0 branch. See if that helps.

Ok :+1: 

----

> It is possible you are doing something I've not thought of and
therefore don't adequately test for it, so the issue is not picked up
with my testing.

I don't think it's possible, my usage is really, really simple. 


> If your able to strip down your code to a minimal version which
reproduces the issue, it would be really handy to get a copy.

Alright, I'll try. Will post today or tomorrow.

**Comment by kamilmielnik** [comment]
Here's a simplified version of my script. I've added a few comments to explain where I get ECONNRESET.

```js
require('cross-fetch/polyfill');
const fs = require('fs');
const path = require('path');
const rimraf = require('rimraf');
const SftpClient = require('ssh2-sftp-client');

const sftpClient = new SftpClient();

const main = async () => {
    try {
        await connect();
    } catch (error) {
        console.error('Cannot establish connection', error);
        process.exit(1);
    }

    ensureDirectoryExists(process.env.DOWNLOAD_DIR);

    let xmlFiles;

    try {
        // This will give an array of a few thousand XML files, I've never got ECONNRESET
        // while listing files.
        xmlFiles = await listXmlFiles(process.env.FTP_ROOT_DIR);
    } catch (error) {
        console.error('Failed to list XML files from FTP', error);
        rimraf.sync(process.env.DOWNLOAD_DIR);
        await disconnect();
        process.exit(1);
    }

    for (let index = xmlFiles.length - 1; index >= 0; --index) {
        const xmlFile = xmlFiles[index];
        try {
            await syncXmlFileData(xmlFile);
        } catch (error) {
            console.error(`Failed to sync data for ${xmlFile.filepath}`, error);

            if (error.code === 'ECONNRESET') {
                await tryToRecoverLostConnection(async () => {
                    try {
                        await syncXmlFileData(xmlFile);
                    } catch (error) {
                        console.error(`Failed to sync data for ${xmlFile.filepath}`, error);
                    }
                });
            }
        }
    }

    rimraf.sync(process.env.DOWNLOAD_DIR);
    try {
        await disconnect();
    } catch (error) {
        console.error('Cannot close connection', error);
        process.exit(1);
    }
    process.exit(0);
};

const connect = async () => {
    console.log('Establishing connection');
    await sftpClient.connect({
        host: process.env.FTP_HOST,
        port: process.env.FTP_PORT,
        username: process.env.FTP_USERNAME,
        password: process.env.FTP_PASSWORD,
        algorithms: {
            cipher: [process.env.FTP_CIPHER],
        },
    });
    console.log('Connection established');
};

const disconnect = async () => {
    console.log('Closing connection');
    await sftpClient.end();
    console.success('Connection closed');
};

const listXmlFiles = async (directoryPath) => {
    console.log(`Listing directory: ${directoryPath}`);

    const xmlFiles = [];
    const fileInfos = await sftpClient.list(directoryPath);

    for (const fileInfo of fileInfos) {
        const { type, name } = fileInfo;
        const filepath = `${directoryPath}/${name}`;

        if (type === 'd') {
            const directoryXmlFiles = await listXmlFiles(filepath);
            xmlFiles.push(...directoryXmlFiles);
        } else {
            if (name.endsWith('.xml')) {
                const xmlFileInfo = { directoryPath, fileInfo, filepath };
                xmlFiles.push(xmlFileInfo);
            }
        }
    }

    console.log(`Listing directory: ${directoryPath} - found ${xmlFiles.length} XML files`);

    return xmlFiles;
};

const tryToRecoverLostConnection = async (action, attempt = 1) => {
    logger.log(`Attempting to re-establish connection in 60s`);
    await delay(60000);

    let connected = false;
    try {
        await connect();
        connected = true;
    } catch (error) {
        console.error('Cannot establish connection', error);
    }

    if (connected) {
        await action();
        return;
    }

    if (attempt < 3) {
        await tryToRecoverLostConnection(action, attempt + 1);
        return;
    }

    const error = new Error('Max number of reconnect attempts has been reached');
    console.error('Cannot establish connection', error);
    throw error;
};

const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

const ensureDirectoryExists = (directoryPath) => {
    if (!fs.existsSync(directoryPath)) {
        fs.mkdirSync(directoryPath);
    }
};

const syncXmlFileData = async (xmlFile) => {
    // ECONNRESET always happens while this function is being executed
    // there are 2 async actions here during which I get ECONNRESET:
    // the `download` and the `fetch`
    const localXmlFilepath = await download(xmlFile.filepath);
    const data = await parseXml(localXmlFilepath);
    rimraf.sync(localXmlFilepath);
    console.log('Syncing data');
    // The server which I'm making a request to here is in no way related to the FTP server
    // where XML files come from. Of course I've simplified the fetch call for the purpose of this example.
    await fetch('https://example.com', {
        method: 'PUT',
        headers: {
            Accept: 'application/json',
        },
        body: JSON.stringify(data),
    });
    console.log('Data synced');
};

const download = async (filepath) => {
    console.log(`Downloading ${filepath}`);
    const basename = path.basename(filepath);
    const targetFilepath = path.join(process.env.DOWNLOAD_DIR, basename);
    await sftpClient.fastGet(filepath, targetFilepath);
    console.log(`Downloading ${filepath} - complete`);
    return targetFilepath;
};

const parseXml = (filepath) => {
    return {
        data1: "I'll save you implementation of this function",
        data2: "it reads an XML file, parses it, and returns an object with extracted data",
    };
};

main();

```

**Comment by theophilusx** [comment]
>> 1. The error event fired before fastGet() had setup the Promise and
> associated error handler. This would result in the 'default' error
> handler being called, which cannot be caught by try/catch.
>
> But then why would the script exit?
>

That I cannot answer. The module itself does not call exit or do
anything to exit the script. All the default error handler does is
report the error to console.log i.e. the Error event .... message. All
it really does is prevent the error event from being unhandled and
bubbling up to the top where it was causing an uncaught error
exception. It would not affect what your script was doing (your script
must just be exiting?).

Previously, the event handler was throwing an error, but because this
event occurs outside the context of the current script's execution path,
there is nothing to catch the error and you get the uncaught exception
error we saw previously (the same reason try/catch cannot work). 

BTW the second part of the output which states the connection has been
closed unexpectedly is from the close event handler. This is the one
which sets client.sftp to undefined and prints the message. It is
setting the client.sftp property to undefined which flags to the system
there is no active sftp connection - this is what all the methods check
for before they execute and why any attempt to perform another operation
after that handler has fired will result in failure. 

The script doesn't exit with anything other than the default 0 exit code
as there is no direct way to do this. I could have the error handler set
an exit code which could be returned by the end() call if that would be
of any use?

Tomorrow I'll see if I can force the sftp.client object to emit a 'fake'
error event. Maybe if I put that on a timer, I may be able to simulate
the error you are seeing. This might help diagnose further.

**Comment by kamilmielnik** [comment]
> I could have the error handler set
an exit code which could be returned by the end() call if that would be
of any use?

I don't think so.

I only wish this was possible:
```js
await sftpClient.fastGet(filepath, targetFilepath); // let's say this succeeds
processFile(targetFilepath); // let's say ECONNRESET happens during this call, which does not use sftpClient instance
try {
    // I'd expect the next call to sftpClient.fastGet (or any other method)
    // to fail with ECONNRESET which is "cached" within sftpClient, just waiting to be thrown
    await sftpClient.fastGet(filepath, targetFilepath);
} catch (error) {
    console.log(error.code === 'ECONNRESET');
}
```

**Comment by theophilusx** [comment]
> I only wish this was possible:
> ```js
> await sftpClient.fastGet(filepath, targetFilepath); // let's say this succeeds
> processFile(targetFilepath); // let's say ECONNRESET happens during this call, which does not use sftpClient instance
> try {
>     // I'd expect the next call to sftpClient.fastGet (or any other method)
>     // to fail with ECONNRESET which is "cached" within sftpClient, just waiting to be thrown
>     await sftpClient.fastGet(filepath, targetFilepath);
> } catch (error) {
>     console.log(error.code === 'ECONNRESET');
> }
> ```

Well that is almost how it is supposed to work :-). The difference is
that instead of getting an econnreset error message, you will get an
ERR_NOT_CONNECTED error when you try to call any sftp methods (except
end()).

Previously, when the error handler was run, it would set SftpClient.sftp
to undefined and then re-throw the error, which resulted in the uncaught
error exception. Now it simply prints a log message. The close event,
which always fires when the connection is reset or closed, now sets the
SftpClient.sftp property to undefined. All methods check SftpClient.sftp
and if it is not defined, throw the ERR_NOT_CONNECTED error.

The reason I don't just cache any errors which are raised via the error
event and then just throw them whenever another request is made is that
I don't know if all the errors raised in this manner are necessarily
connection (or permanent) errors. It seemed cleaner to just flag the
error via console.log and then, if the close event fires as a result of
that error, set SftpClient.sftp to undefined and then throw a bad
connection error if something tries to use it again.

**Comment by theophilusx** [comment]
I see nothing unusual or problematic in your example script. My only
comment would be to be wary of checking specifically for ECONNRESET
errors - that is only one error type which can result in loss of
connection. There are other errors which can occur that will also result
in the connection being lost/closed.

All of the SftpClient methods perform a check for a defined
SftpClient.sftp property before they try to use the connection. If this
property is not defined, an error is thrown (or the promise is rejected)
with the error

{
  message: "No SFTP connection available",
  code: 'ERR_NOT_CONNECTED'
}

**Comment by theophilusx** [comment]
FYI I added the script tools/close-event.js which just tries to do multiple downloads of a file, but with a timer set to emit a fake error and close event. The script works as expected i.e. once the close event fires, any further attempt to do a fastget() fails with the no sftp connection available error. this error is also caught in the catch block of the try/catch. 

> ➜  tools git:(release-5.0.0) ✗ node close-event.js /home/tim/access-s.nc /tmp 50
> Downloading file /home/tim/access-s.nc 50 times
> Downloading to /tmp/test-file.0
> Error event: a fake test error
> sftp: Connection closed unexpectedly
> /tmp/test-file.0 downloaded
> Downloading to /tmp/test-file.1
> Error raised: fastGet: No SFTP connection available
>

**Comment by kamilmielnik** [comment]
> I've updated the release-5.0.0 branch. See if that helps.

It didn't. Today I got ECONNRESET again, and the script just stopped after that. It happened during `fastGet`.

**Comment by theophilusx** [comment]
hate to say it, but I'm out of ideas.

I cannot reproduce the behaviour you are seeing. At the very least, your
scirpt should have got an error when it attempted the next fastGet (or
any other SftpClient method call (except end()).

I don't really understand how your script can just 'stop'. When you say
it stopped, do you mean it just ended or do you mean it appears to have
stopped (hang)?

It is probably time to cut losses. My recommendation would be to not use
the module and instead write your own using SSH2 directly. I don't think
there is anything else I can do to help.

**Comment by theophilusx** [comment]
As the original issue (script exiting due to uncaught exception) has been resolved and I cannot reproduce the other issues, I am going to close this issue. Will release version 5.0.0 later today.
